docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+should also be uninstalled at this point (to support such usage,
+`MCT` should expose instances of any installed plugins that are
+considered optional.)
+
+For example:
+
+```
+define(['mct', './plugins'], function (mct, plugins) {
+  var MCT = mct.MCT;
+
+  function Variant() {
+    MCT.call(this);
+
+    this.install(new plugins.SomePlugin());
+    this.install(new plugins.SomeOtherPlugin(this.core));
+
+    this.uninstall(this.plugins.persistence.localStorage);
+  }
+
+  Variant.prototype = Object.create(MCT.prototype);
+
+  return Variant;
+});
+```
+
+Running an application build using Open MCT then typically looks like:
+
+```
+define(['./Variant], function (Variant) {
+  new Variant().run();
+});
+```
+
+## Writing Plugins
+
+A plugin for Open MCT should inherit from the `Plugin` base class.
+
+Plugins will typically use extension points exposed by other plugins;
+put another way, plugins will typically _depend_ upon other plugins.
+Consistent with "dependency injection as a code style," the preferred
+way for plugins to acquire these references is via constructor
+arguments. (Put another way, in order to use a plugin, you are expected
+to supply its dependencies.) Note that this is not a requirement,
+as Open MCT only ever interacts directly with plugin _instances_;
+any other way of assembling an object with the `Plugin` interface
+should be compatible.
+
+Plugins will also typically expose extension points. The preferred
+way to do this is to expose `Provider` instances as public fields
+of plugins, but this is a matter of code style, and is not enforced
+or expected by the Registration API.
+
+For example, if a plugin depends on the `core` plugin of MCT:
+
+```
+define(['mct', './SomeAction'], function (mct, SomeAction) {
+  var Plugin = mct.Plugin,
+    ServiceProvider = mct.ServiceProvider;
+
+  function ExamplePlugin(core) {
+    Plugin.call(this, function () {
+      core.actionRegistry.register(function () {
+        return new SomeAction();
+      });
+    });
+
+    this.someServiceProvider = new ServiceProvider();
+    this.someServiceProvider.register(function () {
+      return new SomeService();
+    });
+  }
+
+  ExamplePlugin.prototype = Object.create(Plugin.prototype);
+
+  return ExamplePlugin;
+});
+```
+
+Using this plugin then looks like:
+
+```
+define([
+  'mct',
+  './ExamplePlugin',
+  './OtherPlugin'
+], function (mct, ExamplePlugin, OtherPlugin) {
+  var MCT = mct.MCT;
+
+  function MyApplication() {
+    MCT.call(this);
+
+    this.examplePlugin = new ExamplePlugin(this.core);
+    this.otherPlugin = new OtherPlugin(this.examplePlugin);
+
+    this.install(this.examplePlugin);
+    this.install(this.otherPlugin);
+  }
+
+  MyApplication.prototype = Object.create(MCT.prototype);
+
+  return MyApplication;
+});
+```
+
+## Using Extension Points
+
+The services and extensions exposed by providers are retrieved via
+`get` calls to those providers. These calls are expected to occur
```

Comment:
I think this will add a little testing overhead. In addition to mocking the object itself, we will also need to mock the provider.get function. Not a big deal, but it's worth considering the testing implications of some of these changes.